### DISCUSSION TITLE
Can ORM object attributes be computed by a TextClause at add time

### DISCUSSION THREAD

**Discussion by ArsenArsen** [sqlalchemy/sqlalchemy#11379] [Usage Questions] [answered]
Hi!

For very unfortunate reasons, I have some complex ID assigning logic implemented in SQL.
Can this logic be added to a transient ORM object just before a `Session#add`?
This roughly looks as follows:

```py
obj = ORMObject()
obj.obj_id = text("1 + 2")
db.session.add(obj)
```

The above appears to work, but fails to type-check: `error: Incompatible types in assignment (expression has type "TextClause", variable has type "SQLCoreOperations[int] | int")  [assignment]`.

Is this supported?

TIA

PS: I realized just a bit ago that I can probably use the query builder rather than a `text()` clause, and that might type-check, but when I originally asked on Gerrit, I was asked to post a discussion also.  Will try to query-build now.  (EDIT: could query-build and it type checks.. leaving discussion open just in case as the above seems supported but not properly typed)

**Comment by zzzeek** [comment]
this pattern is fine, to deal w/ the typing use `literal_column()` instead of `text()`, the former is more appropriate for a SQL expression written as textual SQL that resolves to a single value.

**Reply by ArsenArsen** [reply]
heh, perhaps I oversimplified in the example above.

the real usage was more like `text("some_fun(:a, :b)").bindparams(bindparam("a", foo, unique=True), bindparam("b", bar, unique=True))` (which, I then realized, was perfectly possible to implement via the query builder completely without text clauses)

the `bindparams` made it so that I needed a `text` clause (AFAICT, I can't bind into a `literal`)

**Reply by zzzeek** [reply]
if it's just a function then use `func.some_fun("a", "b")`, that's a lot easier

**Reply by ArsenArsen** [reply]
> if it's just a function then use `func.some_fun("a", "b")`, that's a lot easier

yes, I did; I replaced the text clause with a nice query builder invocation that was well-typed, but I've been asked (on gitter) to also open a discussion to evaluate the type error, given that the text clause can't appear to pass the type check but does appear to work

**Reply by zzzeek** [reply, marked as answer]
the text() thing is usually going to work almost anywhere, but for a column expression the `literal_column()` is more appropriate.  but if you have the bound parameters in there, there's not a good pattern for that right now, maybe `text(..).columns(Integer).scalar_subquery()`

**Reply by ArsenArsen** [reply]
this appears to work: `text("").columns(foo=Integer()).scalar_subquery()`

thanks!  (but, yes, I ended up just using `func` in the end)
